From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Thu, 1 Jul 2021 21:36:03 -0400
Subject: [PATCH] Add BlockDropResourcesEvent

Adds an event which allows plugin developers
to easily get the items being dropped by any
block instead of only blocks broken by players.

diff --git a/src/main/java/net/minecraft/world/level/block/Block.java b/src/main/java/net/minecraft/world/level/block/Block.java
index 9522e646529f3d849471931b4b3c0d133e7fcfc5..cbdcbfed03afbdf194243d4472e69a664dcf71b9 100644
--- a/src/main/java/net/minecraft/world/level/block/Block.java
+++ b/src/main/java/net/minecraft/world/level/block/Block.java
@@ -302,7 +302,7 @@ public class Block extends BlockBehaviour implements ItemLike {
 
     public static void dropResources(BlockState state, Level world, BlockPos pos) {
         if (world instanceof ServerLevel) {
-            Block.getDrops(state, (ServerLevel) world, pos, (BlockEntity) null).forEach((itemstack) -> {
+            org.bukkit.craftbukkit.event.CraftEventFactory.callBlockDropResourcesEvent(world, pos, Block.getDrops(state, (ServerLevel) world, pos, (BlockEntity) null)).forEach((itemstack) -> { // Parchment
                 Block.popResource(world, pos, itemstack);
             });
             state.spawnAfterBreak((ServerLevel) world, pos, ItemStack.EMPTY, true);
@@ -312,7 +312,7 @@ public class Block extends BlockBehaviour implements ItemLike {
 
     public static void dropResources(BlockState state, LevelAccessor world, BlockPos pos, @Nullable BlockEntity blockEntity) {
         if (world instanceof ServerLevel) {
-            Block.getDrops(state, (ServerLevel) world, pos, blockEntity).forEach((itemstack) -> {
+            org.bukkit.craftbukkit.event.CraftEventFactory.callBlockDropResourcesEvent(world, pos, Block.getDrops(state, (ServerLevel) world, pos, blockEntity)).forEach((itemstack) -> { // Parchment
                 Block.popResource((ServerLevel) world, pos, itemstack);
             });
             state.spawnAfterBreak((ServerLevel) world, pos, ItemStack.EMPTY, true);
@@ -339,7 +339,7 @@ public class Block extends BlockBehaviour implements ItemLike {
 
     public static void dropResources(BlockState state, Level world, BlockPos pos, @Nullable BlockEntity blockEntity, @Nullable Entity entity, ItemStack tool) {
         if (world instanceof ServerLevel) {
-            Block.getDrops(state, (ServerLevel) world, pos, blockEntity, entity, tool).forEach((itemstack1) -> {
+            org.bukkit.craftbukkit.event.CraftEventFactory.callBlockDropResourcesEvent(world, pos, Block.getDrops(state, (ServerLevel) world, pos, blockEntity, entity, tool)).forEach((itemstack1) -> { // Parchment
                 Block.popResource(world, pos, itemstack1);
             });
             state.spawnAfterBreak((ServerLevel) world, pos, tool, true);
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 7b26e9feef965dd0b94658267f4ca8d3a51dceb8..20d7d44288f65e0fd4312701e8e4c6099cb3683e 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -2107,5 +2107,19 @@ public class CraftEventFactory {
             }
         });
     }
+
+    public static List<ItemStack> callBlockDropResourcesEvent(LevelAccessor world, BlockPos pos, List<ItemStack> items) {
+        List<org.bukkit.inventory.ItemStack> bukkitItems = new ArrayList<>(items.size());
+        for (ItemStack item : items) {
+            bukkitItems.add(CraftItemStack.asBukkitCopy(item));
+        }
+        gg.projecteden.parchment.event.block.BlockDropResourcesEvent event = new gg.projecteden.parchment.event.block.BlockDropResourcesEvent(CraftBlock.at(world, pos), bukkitItems);
+        event.callEvent();
+        items = new ArrayList<>(bukkitItems.size());
+        for (org.bukkit.inventory.ItemStack item : bukkitItems) {
+            items.add(CraftItemStack.asNMSCopy(item));
+        }
+        return items;
+    }
     // Parchment end
 }
